client/blocks/keyring-connect-button/index.js

```diff
@@ -0,0 +1,239 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import { last, isEqual, some } from 'lodash';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import {
+	deleteStoredKeyringConnection,
+	requestKeyringConnections,
+} from 'state/sharing/keyring/actions';
+import { getKeyringServiceByName } from 'state/sharing/services/selectors';
+import QueryKeyringServices from 'components/data/query-keyring-services';
+import requestExternalAccess from 'lib/sharing';
+import { getKeyringConnectionsByName } from 'state/sharing/keyring/selectors';
+import { isFetchingConnections } from 'state/sharing/publicize/selectors';
+import { getSelectedSiteId } from 'state/ui/selectors';
+
+class KeyringConnectButton extends Component {
+	static propTypes = {
+		service: PropTypes.oneOfType( [ PropTypes.object, PropTypes.bool ] ),
+		isFetching: PropTypes.bool,
+		keyringConnections: PropTypes.array,
+		siteId: PropTypes.number,
+		onClick: PropTypes.func,
+		onConnect: PropTypes.func,
+	};
+
+	static defaultProps = {
+		onClick: () => {},
+		onConnect: () => {},
+	};
+
+	constructor() {
+		super( ...arguments );
+
+		this.state = {
+			isOpen: false, // The service is visually opened
+			isConnecting: false, // A pending connection is awaiting authorization
+			isDisconnecting: false, // A pending disconnection is awaiting completion
+			isRefreshing: false, // A pending refresh is awaiting completion
+			isAwaitingConnections: false, // Waiting for Keyring Connections request to finish
+		};
+	}
+
+	onClick = () => {
+		this.performAction();
+		this.props.onClick();
+	};
+
+	/**
+	 * Given a service name and optional site ID, returns the current status of the
+	 * service's connection.
+	 *
+	 * @param {string} service The name of the service to check
+	 * @return {string} Connection status.
+	 */
+	getConnectionStatus( service ) {
+		let status;
+
+		if ( this.props.isFetching ) {
+			// When connections are still loading, we don't know the status
+			status = 'unknown';
+		} else if ( ! some( this.props.keyringConnections, { service } ) ) {
+			// If no connections exist, the service isn't connected
+			status = 'not-connected';
+		} else if ( some( this.props.keyringConnections, { status: 'broken' } ) ) {
```

Comment:
Should we filter here as well by `service`?

```js
some( this.props.keyringConnections, { service, status: 'broken' } )
```

Reply:
The answer is - no.
`this.props.keyringConnections` is already filtered to `service.ID`, this function shouldn't even get `service` parameter.